### PR TITLE
Improve "How-to document tests" section

### PR DESCRIPTION
**Pull request by HiddenBug**
This PR aims to improve the how-to document test section in the development_guidlines.md file.

### CODE REVIEW

**Comment by HiddenBug** [comment]
In PR #1 there already started a rather comprehensive discussion on how-to document test properly.
Discussion sequence so far:
- https://github.com/PilzDE/.github/pull/1#discussion_r342401938
- https://github.com/PilzDE/.github/pull/1#discussion_r342638866
- https://github.com/PilzDE/.github/pull/1#discussion_r342943003
Please continue, until problem is resolved...

**Comment by jschleicher** [comment]
Another point: Instead of documenting individual test cases (which should be easily-readable and self-contained), we should start documenting the overall *test concept*. I'd skip this "test-header" completely as it introduces the problems @agutenkunst mentioned. Using doxygen comments as for "normal" code should suffice imho.

**Comment by HiddenBug** [comment]
> Another point: Instead of documenting individual test cases (which should be easily-readable and self-contained), we should start documenting the overall _test concept_. 

Documenting the overall _test concept_ is certainly a good point. I will create a task for this.

> Using doxygen comments as for "normal" code should suffice imho.

I'm not sure why we are suddenly talking about the suddenly called "test-header" as if it is a foreign language. It is in fact simple doxygen. And the first sentence of this suddenly called "test-header", is in fact a short summary of the test intention. So basically nothing else as if documenting "normal" code.

>  I'd skip this "test-header" completely as it introduces the problems @agutenkunst mentioned.

We can of course agree to skip the now so called "test header" to reduce the places which need be created, maintained etc.
I also agree with the suggestion to write  "easily-readable and self-contained" tests. But this proposal is not new. In fact, it should be the case already for all tests. But it simply is not the case. _Why?_ I said it before and I will say it again: **a lack of discipline**. And, furthermore, reviews do often not reveal badly written tests. _Why?_ I don't know. Maybe the reviewer has no time, no motivation, you name it. Much more than a new concept how-to document test, we need something to avoid this lack of discipline (At least this is my opinion).

Other opinions? @ct2034 @leunMar

**Comment by jschleicher** [comment]
Regarding discipline: If it were that simple, you could remedy all principles of fault-tolerance and defensive programming. Please assume that comments that can contradict the code someday will contradict the implementation. Not because someone isn't a good programmer, but "because it works" and there are time constraints and redundant complexity (doing duplicate things is not what humans are used to). So yes: Describing the purpose of the test function in doxygen at a high level is unquestioned. But the steps could vanish from the comment in my opinion.

One more aspect: Some time ago I've read an article propagating spaghetti-code for tests. The author reasoned, that re-usablility is less of interest for tests but self-contained (maybey copied over and over) tests are worth breaking some concepts of "good code".

**Comment by HiddenBug** [comment]
> One more aspect: Some time ago I've read an article propagating spaghetti-code for tests. The author reasoned, that re-usablility is less of interest for tests but self-contained (maybey copied over and over) tests are worth breaking some concepts of "good code" for tests.

I am really sorry, but for the sake of clarity, I cannot leave your statement without comment.

If anybody asks: The first answer to your generalized statement "[...] tests are worth breaking some concepts of "good code" [...]" always has to be **NO**. That is something, I think, should also be stated in the developement guidlines. As a good starting point for any discussion about CleanCode and tests, I consider the musing of Robert C. Martin in "Clean Code - A Handbook of Agile Software Craftsmanship" at the beginning of chapter 9 "Unit Tests". He describes what happens if you start neglecting the clean code rules.

It was already mentioned by @agutenkunst that there are some discussions about DAMP vs. DRY (see for example on [StackOverflow](https://stackoverflow.com/questions/6453235/what-does-damp-not-dry-mean-when-talking-about-unit-tests/11837973#11837973)) when it comes to tests. And I'm totally open for suggestions what kind of approach we want to choose for our tests. But I also want to state that there are some strict limits, especially when it comes to re-usability considerations in tests. I think the best example is our [AsynTest class](https://github.com/PilzDE/pilz_robots/blob/melodic-devel/pilz_testutils/include/pilz_testutils/async_test.h). 
At first, everybody did what he considered right, when it came to threading in tests: One person used listeners, another one simply used sleep and again another person simply copied things from other tests. But as time marched on, we started having more and more problems with threaded tests and even more troubling: These tests were difficult to read and maintain. In the end, one of us, sat down and wrote a helper class, in other words, considered the re-usability. And as a consequence the tests became significantly more easily to read and maintain. Because somebody sat down and thought hard (and maybe long - who knows) about how to name things, wrote a documentation on how-to use the helper class and, finally, unified all previously existing tests. And as a side effect: Suddenly everybody knew what to do and use in case of threaded tests.
My point is: Duplicating things in tests can be of advantages when it comes to readability. But never should we state in general that "[...] re-usability is [of] less [...] interest for tests" or that "[...] tests are worth breaking some concepts of "good code" [...]". There are a few exceptions when it comes to CleanCode and tests but they should be clearly stated in the development guidelines and used in a well defined and more importantly limited environment.

Please also state the article if possible, so that everybody can follow the arguments of the original author. Thanks.

File: development_guidelines.md [hunk 1]
```diff
@@ -212,6 +212,30 @@ is used:
...
+
+Expected Results:
+  0. Explain in one short sentence what result you expect in step 1.
+  1. Explain in one short sentence what result you expect in step 2.
+  ...
+
+``` 
 For a proper test documentation please do not forget to link the test against
 the corresponding requirement. On how-to do that see the
 [requirement linking section](#how-to-link-a-requirement-against-a-test).
```
Comment (leunMar) [inline development_guidelines.md]: Can be removed? See new lines 221, 222

**Comment by jschleicher** [comment]
Yep, the article probably was about [DAMP vs DRY](https://stackoverflow.com/questions/6453235/what-does-damp-not-dry-mean-when-talking-about-unit-tests) allowing redundancy in a test case in favor of readability. Since my browser bookmarks are lost, I'm currently unable to find the link to the original article :-(

**Comment by leunMar** [comment]
So this is about test documentation, right?
I think the guiding principle should be to strive for self-explanatory test naming/code and only add documentation (apart from the requirement-linking) if the test is otherwise hard to understand. This documentation would be optional (however a reviewer could require it) and there would be no general rules for it (like the current step-by-step description).
So I suggest a guiding principle (no docu) with the possibility of deviating from it if it makes sense.

**Comment by leunMar** [comment]
After all, this should be a guideline and not a summary of rules. It should help the developer to know what he should care about. It should not be too restrictive.

**Comment by ct2034** [comment]
@hslusarek Can you address the review by @leunMar and then merge the PR?
If required we can open an issue for https://github.com/PilzDE/.github/pull/3#issuecomment-551509459

**Comment by hslusarek** [comment]
@martiniil Can you please review the changes?

**Review by martiniil — CHANGES_REQUESTED** [review]
I found a small grammar issue. You can merge after commiting the change.

**Pull request merged**